### PR TITLE
[SPARK-13369] Make number of consecutive fetch failures for a stage b…

### PR DESCRIPTION
**Pull request by sitalkedia**
## What changes were proposed in this pull request?

The previously hardcoded max 4 retries per stage is not suitable for all cluster configurations. Since spark retries a stage at the sign of the first fetch failure, you can easily end up with many stage retries to discover all the failures. In particular, two scenarios this value should change are (1) if there are more than 4 executors per node; in that case, it may take 4 retries to discover the problem with each executor on the node and (2) during cluster maintenance on large clusters, where multiple machines are serviced at once, but you also cannot afford total cluster downtime. By making this value configurable, cluster managers can tune this value to something more appropriate to their cluster configuration.

## How was this patch tested?

Unit tests

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/scheduler/Stage.scala [hunk 1]
```diff
@@ -59,6 +59,7 @@ private[scheduler] abstract class Stage(
     val numTasks: Int,
     val parents: List[Stage],
     val firstJobId: Int,
+    val maxConsecutiveFetchFailures: Int,
```
Comment (kayousterhout) [inline core/src/main/scala/org/apache/spark/scheduler/Stage.scala]: I don't think you need this anymore

File: core/src/main/scala/org/apache/spark/scheduler/ShuffleMapStage.scala [hunk 2]
```diff
@@ -41,9 +41,10 @@ private[spark] class ShuffleMapStage(
     numTasks: Int,
     parents: List[Stage],
     firstJobId: Int,
+    maxConsecutiveFetchFailures: Int,
```
Comment (kayousterhout) [inline core/src/main/scala/org/apache/spark/scheduler/ShuffleMapStage.scala]: I don't think you need this anymore

File: core/src/main/scala/org/apache/spark/scheduler/ResultStage.scala [hunk 3]
```diff
@@ -34,8 +34,10 @@ private[spark] class ResultStage(
     val partitions: Array[Int],
     parents: List[Stage],
     firstJobId: Int,
+    maxConsecutiveFetchFailures: Int,
```
Comment (kayousterhout) [inline core/src/main/scala/org/apache/spark/scheduler/ResultStage.scala]: I don't think you need this anymore.

File: core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala [hunk 4]
```diff
@@ -187,6 +187,11 @@ class DAGScheduler(
   /** If enabled, FetchFailed will not cause stage retry, in order to surface the problem. */
   private val disallowStageRetryForTest = sc.getConf.getBoolean("spark.test.noStageRetry", false)
 
+  /** Number of consecutive fetch failures allowed before a stage is aborted */
```
Comment (kayousterhout) [inline core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala]: Can you change this and the comment on line 1738 to the comment Imran suggested for the docs?

File: docs/configuration.md [hunk 5]
```diff
@@ -1506,6 +1506,11 @@ Apart from these, the following properties are also available, and may be useful
     of this setting is to act as a safety-net to prevent runaway uncancellable tasks from rendering
     an executor unusable.
   </td>
+  <td><code>spark.stage.maxAttempts</code></td>
```
Comment (kayousterhout) [inline docs/configuration.md]: @squito maxConsecutiveAttempts? Realized maxAttempts is slightly misleading / might be good to emphasize that they're consecutive

File: core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala [hunk 6]
```diff
@@ -187,6 +187,11 @@ class DAGScheduler(
...
+  private[scheduler] val maxConsecutiveFetchFailuresPerStage =
```
Comment (kayousterhout) [inline core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala]: Ok one more naming nit: can you make this consistent with the config name, so maxConsecutiveStageAttempts (and ditto with the constant)?

File: core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala (hunk 4 above)
Comment (sitalkedia) [inline core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala]: done

- [ ]

File: core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala (hunk 6 above)
Comment (sitalkedia) [inline core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala]: sure, done

File: core/src/main/scala/org/apache/spark/scheduler/ResultStage.scala (hunk 3 above)
Comment (sitalkedia) [inline core/src/main/scala/org/apache/spark/scheduler/ResultStage.scala]: Good point.

File: docs/configuration.md (hunk 5 above)
Comment (sitalkedia) [inline docs/configuration.md]: +1

File: docs/configuration.md [hunk 7]
```diff
@@ -1506,6 +1506,11 @@ Apart from these, the following properties are also available, and may be useful
     of this setting is to act as a safety-net to prevent runaway uncancellable tasks from rendering
     an executor unusable.
   </td>
+  <td><code>spark.stage.maxConsecutiveAttempts</code></td>
+  <td>4</td>
+  <td>
+    Number of consecutive stage retries allowed before a stage is aborted (since you can have multiple fetch failures in one stage attempt)
```
Comment (kayousterhout) [inline docs/configuration.md]: Oops sorry just noticing that I don't think @squito intended for the part in parens to be in the comment?  You could make this instead 

Number of consecutive stage retries allowed before a stage is aborted (stages are retried if shuffle fetch failures occur).

or, you could just eliminate the part in parens.  Sorry I didn't notice this earlier!

Comment (sitalkedia) [inline docs/configuration.md]: Okay, removed the part in parens.

**Comment by sitalkedia** [comment]
Jenkins retest this please.

**Comment by kayousterhout** [comment]
@sitalkedia please do not ignore flaky test failures.  I updated the JIRA associated with this one: https://issues.apache.org/jira/browse/SPARK-19803?jql=project%20%3D%20SPARK%20AND%20labels%20%3D%20flaky-test but please be sure to do this yourself before re-testing next time.

File: docs/configuration.md [hunk 8]
```diff
@@ -1506,6 +1506,11 @@ Apart from these, the following properties are also available, and may be useful
     of this setting is to act as a safety-net to prevent runaway uncancellable tasks from rendering
     an executor unusable.
   </td>
+  <td><code>spark.stage.maxConsecutiveAttempts</code></td>
+  <td>4</td>
+  <td>
+    Number of consecutive stage retries allowed before a stage is aborted.
```
Comment (squito) [inline docs/configuration.md]: there is a off-by-one difference between "attempts" and "retries" -- eg. if this is set to 1, do you allow one retry, or do you give up after one attempt?  I realize this is super minor but I remember dealing with confusion about this for task failures.  I don't think which one we use matters a ton, but the implementation here is "attempt", so how about just rewording the doc to "Number of consecutive stage attempts ...".

Same goes for the comments which use "retries".

(I see now this was wording was my fault, from making one suggestion in one place, and another elsewhere, sorry about that.)

**Review by squito — CHANGES_REQUESTED** [review]
just a minor detail I'd like updated in the docs, otherwise lgtm

File: docs/configuration.md (hunk 8 above)
Comment (kayousterhout) [inline docs/configuration.md]: Hah sorry for all of the comment changes from the combination of Imran and me!! But I agree that this was an issue before and would be good to update.  Thanks for the many updates here @sitalkedia.

Comment (sitalkedia) [inline docs/configuration.md]: That makes sense, updated accordingly. Thanks for the quick review.

**Comment by sitalkedia** [comment]
The tests passes on my laptop, so looks like we have more flaky tests?

**Comment by squito** [comment]
yeah, sorry I am looking, but keep getting distracted ... I'm sure these failures don't matter but can't merge this second anyhow so lets just test again ...

**Comment by squito** [comment]
Jenkins, retest this please

**Comment by kayousterhout** [comment]
@squito FYI I filed a JIRA for the 2nd of the two unit tests that failed in that run (looks like you'd already filed a JIRA for the first one)

**Comment by kayousterhout** [comment]
Jenkins, retest this please

**Comment by squito** [comment]
merged to master (slightly reworded the first line of the commit msg so it all fit).

thanks @sitalkedia, especially for sticking with this despite the delays, our nitpickiness, and the flaky tests

**Pull request closed without merging**